### ISSUE TITLE
Feature request: Data Source for Lambda function

### ISSUE THREAD

**Issue by et304383** [terraform-providers/terraform-provider-aws#2578]
The use case is simple: we create Lambda functions via serverless and want to reference that Lambda function in a terraform file when creating a Cognito user pool for custom verification flow.

There appears to be no easy way to get the Lambda function ARN in our terraform file.

**Comment by loivis** [comment]
`aws_lambda_function` by default exports an `arn` attribute. We used it to create cognito user pool and it works well.

**Comment by apparentlymart** [comment]
Hi @et304383! Thanks for this feature request.

We are generally in favor of adding new data sources where it makes it possible to fetch additional attributes for an object that wouldn't otherwise naturally be available, as in this case.

The Terraform team at HashiCorp won't be able to work on this in the near future due to our focus being elsewhere, but we'd be happy to review a pull request if you or someone else has the time and motivation to implement it. Alternatively, if others would also like to see this implemented I'd encourage adding a :+1: upvote reaction to the original issue comment (not to _this_ comment), which we use as one of the inputs to prioritize work for the Terraform team.

**Comment by et304383** [comment]
@loivis I explicitly said the Lambda fucntion is created outside of Terraform via serverless.

**Comment by jammycakes-ukpds** [comment]
As a workaround in the meantime, the ARN of a Lambda function is deterministic:

    arn:aws:lambda:<region>:<account_id>:function:<function_name>

You can therefore do something like this:

```
data "aws_caller_identity" "current" {}

data "aws_region" "current" {}
```

Then reference your lambda function as:

```
"arn:aws:lambda:${data.aws_caller_identity.current.account_id}:${data.aws_region.current.name}:function:my_function"
```

**Comment by soulrebel** [comment]
@jammycakes-ukpds kudos, but region goes before account. This is the correct syntax:
```
"arn:aws:lambda:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:function:my_function"
```

**Comment by t-pascal** [comment]
Has this been addressed in #2984 ?

**Comment by bflad** [comment]
Thanks @t-pascal! You're right. We have introduced a new `aws_lambda_function` data source via #2984 and it will release with v1.16.0 of the AWS provider, likely in a few days. 🎉

**Comment by bflad** [comment]
This has been released in [version 1.16.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1160-april-25-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
